### PR TITLE
split code path for glCopyImageSubDataEXT&glCopyImageSubData

### PR DESCRIPTION
**Pull request by qiankanglai**
This changes splits glCopyImageSubDataEXT&glCopyImageSubData (rather than aliasing).

Angle's Vulkan backend has already supported `glCopyImageSubDataEXT` and it works fine, but we cannot call `glCopyImageSubData` instead. Because currently Angle's max supported version is 3.1, and `glCopyImageSubData` validate context for 3.2. (which means we cannot find a way to take use the non-ext API for now).

So with RenderDoc attached, our code fails because call to `glCopyImageSubDataEXT`  is forced into `glCopyImageSubData` .

### CODE REVIEW

**Comment by baldurk** [comment]
This is not the right fix in my view, there is no reason to split this function in particular and not others. It is very common in GL for an extension function to be promoted but with the promoted version possibly not available on all drivers. The fetching of function pointers should handle this such that cases where an extension function isn't available.

Can you explain in more detail what is going wrong? If angle only supports `glCopyImageSubDataEXT` and not `glCopyImageSubData` that should be fine, the EXT version of the function should be fetched and used. I don't understand your explanation of what is wrong exactly or why only this function is affected.

As a side note, in my experience ANGLE is not very reliable and its GL implementation is not sufficiently complete for RenderDoc's replay, so I strongly recommend using GL directly with a real driver and not using RenderDoc with ANGLE or you may encounter other problems.

**Comment by qiankanglai** [comment]
Yep, I have to admit this change is a bit strange. Here is the detail case: Angle has both [GL_CopyImageSubData](https://github.com/google/angle/blob/2b8172b9b21edeeb0228dfe1dc1640447864522c/src/libGLESv2/entry_points_gles_3_2_autogen.cpp#L178) and [GL_CopyImageSubDataEXT](https://github.com/google/angle/blob/2b8172b9b21edeeb0228dfe1dc1640447864522c/src/libGLESv2/entry_points_gles_ext_autogen.cpp#L4453). They share almost the same code path except validation part below:

[ValidateCopyImageSubData](https://github.com/google/angle/blob/2b8172b9b21edeeb0228dfe1dc1640447864522c/src/libANGLE/validationES32.cpp#L136)
```
if (context->getClientVersion() < ES_3_2)
{
    context->validationError(entryPoint, GL_INVALID_OPERATION, kES32Required);
    return false;
}
```

[ValidateCopyImageSubDataEXT](https://github.com/google/angle/blob/2b8172b9b21edeeb0228dfe1dc1640447864522c/src/libANGLE/validationESEXT.cpp#L896)
```
if (!context->getExtensions().copyImageEXT)
{
    context->validationError(entryPoint, GL_INVALID_OPERATION, kExtensionNotEnabled);
    return false;
}
```

At the same time, [RendererVk::getMaxSupportedESVersion](https://github.com/google/angle/blob/2b8172b9b21edeeb0228dfe1dc1640447864522c/src/libANGLE/renderer/vulkan/RendererVk.cpp#L2792) limits the max version to 3.1, which makes `ValidateCopyImageSubData` always fail 😞 

> ANGLE is not very reliable and its GL implementation is not sufficiently complete

Yep I agree with that but debugging on PC with RenderDoc & GLES is much more convenient than on Android. So we employ this method as the first round when encounting problems, and walk into Android later. (BTW Angle's Vulkan Backend is much better than others)

Any suggestion? I'm also considering yesterday modifying Angle (just skip the annoying validation locally, because it is for in house development rather than publish out) last night.

**Comment by baldurk** [comment]
Unless I'm missing something then from what you're describing this sounds like an ANGLE issue, if it has functions that it doesn't support then I wouldn't expect it to return those function pointers. If it returned NULL then RenderDoc would fall back to using `glCopyImageSubDataEXT` (or if that isn't available either, fall back to emulating them).

FWIW if you're debugging on PC I would suggest just using a native PC OpenGL driver. GLES is more-or-less just a broken & worse  version of OpenGL, so having a codepath that uses a subset of OpenGL will still let you test on a proper platform with a good GL driver. Since most of the difficulties in porting in my experience come from driver/device-specific bugs and quirks, ANGLE likely isn't going to replicate any better than a PC driver. Note that if you do use ANGLE you may find that RenderDoc breaks on replay (that's what I've found in the past) and may be unreliable on capture depending on the circumstances since layered API-on-API wrappers are not supported.

**Comment by qiankanglai** [comment]
Got it, thanks for replay 👍 ! I'll patch Angle for temporal use and later switch to native PC GL
